### PR TITLE
Remove chance of loop with records that are rolledback

### PR DESCRIPTION
**Pull request by krainboltgreene**
Fix for #14

### CODE REVIEW

File: lib/activerecord/delay_touching/state.rb [hunk 1]
```diff
@@ -49,6 +49,16 @@ def clear_records
         @already_updated_records.clear
       end
 
+      # If we don't do this then an infinite loop is possible due to how 
+      # Set#subtract and ActiveRecord::Core#== work with the in memory changes 
+      # from a rollback and active record sync
+      def remove_unpersisted_records!
+        @records.each do |_, set|
```
Comment (mtuckergd) [inline lib/activerecord/delay_touching/state.rb]: Need to change `_` back to `attr` sicne you added the delete on line 59

**Comment by krainboltgreene** [comment]
@mtuckergd Can you commit the test to my branch?

File: lib/activerecord/delay_touching/state.rb [hunk 2]
```diff
@@ -49,6 +49,16 @@ def clear_records
         @already_updated_records.clear
       end
 
+      # If we don't do this then an infinite loop is possible due to how
```
Comment (oehlschl) [inline lib/activerecord/delay_touching/state.rb]: FWIW, if you are going to comment something, I'd mention that ActiveRecord lazily syncs transaction state after rollback, which my change in-memory state of key objects, which requires that the hash be rekeyed.

Set#subtract and ActiveRecord::Core#== work as expected (and aren't worth mentioning, IMO); it's really the lazy updating on rollbacks that's the kicker.

File: lib/activerecord/delay_touching/state.rb [hunk 3]
```diff
@@ -49,6 +49,16 @@ def clear_records
         @already_updated_records.clear
       end
 
+      # If we don't do this then an infinite loop is possible due to how 
+      # Set#subtract and ActiveRecord::Core#== work with the in memory changes 
+      # from a rollback and active record sync
+      def remove_unpersisted_records!
+        @records.each do |attr, set|
+          set.instance_variable_get(:@hash).rehash
```
Comment (oehlschl) [inline lib/activerecord/delay_touching/state.rb]: (Same as what I commented on your local PR.)

I think for rehashing on L57 to have an effect, it relies on ActiveRecord's sync_with_transaction_state being invoked upstream. From the AR source, that method is lazily executed, and if it's not run, I think rehash will have no effect. (See below.)

If that is the case, I think it's clearer to invoke explicitly here (as suggested in #14), whether it's via the private method itself or some public API.

For example:
```ruby
# SETUP

records = Set.new
ActiveRecord::Base.transaction do
  records << Micropost.create!(post_type: 'topic', post_category: 'log', content: 'test')
  raise ActiveRecord::Rollback
end

# BAD

records.instance_variable_get(:@hash).rehash # micropost still has ID set (no effect)
records.keep_if(&:persisted?) # this keeps the micropost above (since key is out of sync)

# GOOD

records.each {|record| record.send(:sync_with_transaction_state) }
records.instance_variable_get(:@hash).rehash # now rehashing has an effect
records.keep_if(&:persisted?)  # works as expected
```

If this does work, I think it's only incidental, and that something more explicit would be better.

Comment (krainboltgreene) [reply lib/activerecord/delay_touching/state.rb]: Here's the example I'm working with:

```
ActiveRecord::Base.delay_touching do
  ActiveRecord::Base.transaction do
    # write and touch any new record
    a = Order.create!
    a.touch

    b = Order.create!

    User.last.touch
    Order.first.touch
    raise ActiveRecord::Rollback
  end

  # results in an infinite loop
end
```

The result should be:

  - [x] No infinite loop
  - [x] A single touch update to the last user
  - [x] A single touch update to the first order

So far so good. Here are the logs:


```
(0.3ms)  BEGIN

SQL (0.5ms)  INSERT INTO "orders" ("status", "number", "created_at", "updated_at") VALUES ($1, $2, $3, $4) RETURNING "id"  [["status", "pending"], ["number", 27525204], ["created_at", "2018-08-13 20:06:23.381004"], ["updated_at", "2018-08-13 20:06:23.381004"]]

Order Exists (0.5ms)  SELECT  1 AS one FROM "orders" WHERE "orders"."number" = 36529264 LIMIT 1

SQL (0.8ms)  INSERT INTO "orders" ("status", "number", "created_at", "updated_at") VALUES ($1, $2, $3, $4) RETURNING "id"  [["status", "pending"], ["number", 36529264], ["created_at", "2018-08-13 20:06:23.404695"], ["updated_at", "2018-08-13 20:06:23.404695"]]

User Load (1.2ms)  SELECT  "users".* FROM "users"  ORDER BY "users"."id" DESC LIMIT 1

Order Load (1.0ms)  SELECT  "orders".* FROM "orders"  ORDER BY "orders"."id" DESC LIMIT 1

(0.3ms)  ROLLBACK

remove_unpersisted_records

ATTR

SET #<Set: {#<Order id: nil, number: 27525204>, #<User id: 5335560>, #<Order id: 31426539, number: 36529264>}>

REHASH {#<Order id: nil, number: 27525204>=>true, #<User id: 5335560>=>true, #<Order id: 31426539, number: 36529264>=>true}

STRIP #<Set: {#<User id: 5335560>, #<Order id: 31426539, number: 36529264>}>

EMPTY? false

(0.3ms)  BEGIN

SQL (0.9ms)  UPDATE "users" SET "updated_at" = '2018-08-13 20:06:23.440097' WHERE "users"."id" = 5335560

SQL (0.6ms)  UPDATE "orders" SET "updated_at" = '2018-08-13 20:06:23.443024' WHERE "orders"."id" = 31426539

(0.6ms)  COMMIT
```

**Comment by mtuckergd** [comment]
@krainboltgreene Per your request, I pushed the infinite loop spec. I also included a more granular version requirement for Rails (since specs fail in 5.x, so this lib likely doesn't work there) and upgraded rspec. The fix certainly does the trick in resolving the spec failure, at least. 

I'm looking at the fix a little closer today, and I'm not super comfortable with the approach of tapping into the private hash instance within the `Set` object. This breaks abstraction and may make the library brittle to any internal changes of the Set implementation (including simply renaming the internal variable). I'd like to dig a little more to see if we can accomplish the same result using only public object interfaces.

**Comment by krainboltgreene** [comment]
@mtuckergd Thanks so much for pushing that.

> I'm not super comfortable with the approach of tapping into the private hash instance within the Set object

I understand this concern, but you probably have users that don't have Ruby 2.5 where [`Set#reset`](https://ruby-doc.org/stdlib-2.5.1/libdoc/set/rdoc/Set.html#method-i-reset) is implemented.

File: lib/activerecord/delay_touching/state.rb [hunk 4]
```diff
@@ -49,6 +49,18 @@ def clear_records
...
       end
 
+      # If we don't do this then an infinite loop is possible due to how 
+      # Set#subtract and ActiveRecord::Core#== work internally, ActiveRecord lazily syncs
+      # transaction state after rollback, which may change in-memory state of key objects,
+      # which requires that the hash be rekeyed.
+      def remove_unpersisted_records!
+        @records.each do |attr, set|
+          set.each(&:persisted?)
+          set.instance_variable_get(:@hash).rehash
```
Comment (mtuckergd) [inline lib/activerecord/delay_touching/state.rb]: I'm not super comfortable with this access into private implementation of Set. 

It looks like Ruby 2.5 introduced the `reset` method to handle exactly this situation (see https://bugs.ruby-lang.org/issues/6589) but that doesn't help us in any prior versions :(.

I'm thinking through whether there's a reasonable alternative to this approach. Otherwise, we could potentially introduce a fallback to this mechanism only if `reset` is not defined (thus using public api for ruby 2.5+ and reducing the risk). I'm definitely open to other suggestions, as well.

Comment (oehlschl) [reply lib/activerecord/delay_touching/state.rb]: @mtuckergd @krainboltgreene depending on how far you want to take this, you could just add a private shim to the State class.

```
private

# Set#reset is not introduced until Ruby 2.5
def reset_set(set)
if set.respond_to?(:reset)
   set.reset
else
   set.instance_variable_get(:@hash).rehash
end
```

Comment (mtuckergd) [reply lib/activerecord/delay_touching/state.rb]: Yep, good call. That's along the lines of the fallback I was considering. I have one other thought I want to pursue before we go this route.

Comment (krainboltgreene) [reply lib/activerecord/delay_touching/state.rb]: Sounds great.

**Comment by mtuckergd** [comment]
Okay @krainboltgreene and @oehlschl - I was able to spend some time really wrapping my brain around what's going on here, so I think I've caught up to your thinking. I also don't see a great way around the hack of forcing a rekey of the Set object, unfortunately. I'd like to use the fallback approach we discussed, but add some very clear documentation on why we had to introduce the hack. I'm happy to take what we've discussed and make those changes (I can push to @krainboltgreene's branch for this PR).

Just so I'm clear on the situation and can provide a coherent description in the code comment, here's my understanding of the root cause. Can you let me know if you have a different understanding?

> 
> The Set class identifies object uniqueness using the #hash method on the objects in its collection.
> 
> When a newly created ActiveRecord object is subsequently touched within a delay_touch block, it is tracked in the State class (in a Set object) using a hash that is defined by the assigned primary key ID from its initial save. However, if that transaction gets rolled back, the ID gets cleared from the ActiveRecord object, thus changing its hash value, based on the behavior defined in ActiveRecord::Core#hash.
> ```
> def hash
>   if id
>     id.hash
>   else
>     super
>   end
> end
> ```
> 
> At the end of the transaction, the State operation is unable to clear the ActiveRecord object from the Set object, because its hash value after the rollback has changed, due to no longer having a valid ID (again, see ActiveRecord::Core#hash implementation). This causes delay_touching to repeatedly attempt to touch and remove the record, resulting in an infinite loop.
> 
> The only feasible way to prevent this infinite loop is to force a rehash of the Set object. As of Ruby 2.5, this can be easily done uisng Set#reset, but prior to that version you have to directly access the private @hash instance variable, breaking the abstraction of the Set object and making the solution susceptible to breaking when the Set implementation changes.
> 
> The best recommendation we have is to check if the Set#reset method is available, and fall back to accessing the private hash in the Set object, to rekey the set and allow objects to be removed from the collection.
> 

I will make the code comment much more concise, but want to ensure I have the right understanding of root cause here.

thanks!

**Comment by mtuckergd** [comment]
I just posted my thoughts on two possible approaches to this issue. Feedback welcome!

**Comment by oehlschl** [comment]
@mtuckergd I think your summary is accurate and helpful. FWIW, I tried to clarify a few parts myself; feel free to use any or none of the following:

> TL;DR: This is required to prevent an infinite loop in the case of ActiveRecord rollbacks.
> 
> When an ActiveRecord object is touched within a delay_touch block, it is tracked in the State class and stored inside a Set object. However, in the case that an ActiveRecord object is both created and touched within the same transaction, and that transaction is rolled back, that object should not be eligible for a delayed touch. In other words, even though all touched objects are still stored in the State, we should touch only those that have been actually persisted to the database.
>
> While it sounds simple, that act is complicated by the way ActiveRecord persistence and Sets are implemented under the hood. Set *values* are actually stored as Hash *keys*, so Set membership relies heavily on Hash key identity comparison. Additionally, ActiveRecord lazily unsets the ID attribute from unpersisted ActiveRecord objects in the case of rollbacks, which changes the object’s hash value based on the behavior defined in ActiveRecord::Core#hash:
> 
> https://github.com/rails/rails/blob/v4.2.10/activerecord/lib/active_record/core.rb#L398-L404
> 
> Therefore, after a transaction is rolled back, the State operation is unable to clear any unpersisted ActiveRecord objects from the Set object because their IDs (and therefore their hash values) have changed since they were originally added to the Set (or, “since they were originally keyed in the Hash”).  This causes delay_touching to repeatedly attempt to touch and remove the record, resulting in an infinite loop.
> 
> The only feasible way to prevent this infinite loop is to force a sync of the ActiveRecord transaction state followed by a rehash of the Set object. As of Ruby 2.5, this can be easily done using Set#reset, but prior to that version you have to directly access the private `@hash` instance variable, breaking the abstraction of the Set object and making the solution susceptible to breaking when the Set implementation changes.
> 
> The best recommendation we have is to check if the Set#reset method is available, and fall back to accessing the private hash in the Set object, to rekey the set and allow objects to be removed from the collection.

Thanks!

**Comment by mtuckergd** [comment]
@krainboltgreene @oehlschl I'd love your input on the alternative PRs I posted. I am fairly certain that https://github.com/godaddy/activerecord-delay_touching/pull/18 will be a much more efficient approach, but the commit history on this PR looks like you guys may have looked at approaching in `more_records?` and changed direction, so I want to make sure we're not missing something there.

Appreciate your feedback on those!

**Comment by krainboltgreene** [comment]
I much prefer the #17 but I believe both of them will work after mapping it out on paper. I leave to your discretion.

**Comment by mtuckergd** [comment]
@krainboltgreene can you give me some insights on your preference for 17? I have a few reasons why I believe 18 is a more solid approach, but I could certainly be missing something.

LMK if you prefer to chat directly and I can email you.

**Comment by mtuckergh** [comment]
This is fixed as of https://github.com/godaddy/activerecord-delay_touching/commit/e71ade4ae4280859df8021fa0e56ef79b2ac02fd.
Version 1.1.0 should be available on rubygems sometime today.
